File: VB/MasterDetailExample/MergeProcessor.vb
Imports System.Collections.Generic
Imports DevExpress.XtraRichEdit
Imports DevExpress.XtraRichEdit.API.Native

Namespace MasterDetailExample

    Friend Class MergeProcessor

        Private mainRichEdit As RichEditDocumentServer = New RichEditDocumentServer()

        Private suppllierRichEdit As RichEditDocumentServer = New RichEditDocumentServer()

        Private productRichEdit As RichEditDocumentServer = New RichEditDocumentServer()

        Private ordersRichEdit As RichEditDocumentServer = New RichEditDocumentServer()

        Private resultRichEdit As RichEditDocumentServer = New RichEditDocumentServer()

        Private ds As SupplierCollection

        Private dataDetailedForProducts As ProductCollection

        Private dataDetailedForOrders As OrderDetailCollection

        Private Shared fakeDataSource As List(Of Integer) = DataHelper.CreateFakeDataSource()

        Private supplierID As Integer = -1

        Private productID As Integer = -1

        Public Sub New()
            ' Subscribe to the CalculateDocumentVariable event that triggers the master-detail report generation
            AddHandler resultRichEdit.CalculateDocumentVariable, New CalculateDocumentVariableEventHandler(AddressOf resultRichEdit_CalculateDocumentVariable)
            ' Load main template
            mainRichEdit.LoadDocument("main.rtf")
            ' Create project's data source
            ds = DataHelper.CreateData()
            ' Load templates and specify data sources for RichEdit servers.
            suppllierRichEdit.LoadDocument("supplier.rtf")
            suppllierRichEdit.Options.MailMerge.DataSource = ds
            productRichEdit.LoadDocument("detail.rtf")
            productRichEdit.Options.MailMerge.DataSource = ds
            productRichEdit.Options.MailMerge.DataMember = "Products"
            ordersRichEdit.LoadDocument("detaildetail.rtf")
            ordersRichEdit.Options.MailMerge.DataSource = ds
            ordersRichEdit.Options.MailMerge.DataMember = "Products.OrderDetails"
        End Sub

'#Region "#startmailmerge"
        ' Start the process by merging main template into the document contained within the resultRichEdit server.
        Public Sub Start()
            ' Since the main template contains no merge fields requiring no merge data, provide a fake data source.
            ' Otherwise mailmerge will not start.
            mainRichEdit.Options.MailMerge.DataSource = fakeDataSource
            ' Trigger the multistage process. After the first mailmerge the CalculateDocumentVariable event
            'for the resultRichEdit server fires.
            mainRichEdit.MailMerge(resultRichEdit.Document)
            resultRichEdit.SaveDocument("result.docx", DocumentFormat.OpenXml)
        End Sub

'#End Region  ' #startmailmerge
'#Region "#secondstage"
        ' Second stage. For each Supplier ID create a detailed document that will be inserted in place of the DOCVARIABLE field.
        Private Sub resultRichEdit_CalculateDocumentVariable(ByVal sender As Object, ByVal e As CalculateDocumentVariableEventArgs)
            If Equals(e.VariableName, "Supplier") Then
                ' Create a text engine to process a document after the mail merge.
                Dim richServerMaster As RichEditDocumentServer = New RichEditDocumentServer()
                ' Provide a procedure for further processing
                AddHandler richServerMaster.CalculateDocumentVariable, AddressOf richServerMaster_CalculateDocumentVariable
                ' Create a merged document using the Supplier template. The document will contain DOCVARIABLE fields with ProductID arguments. 
                ' The CalculateDocumentVariable event for the richServerMaster fires.
                suppllierRichEdit.MailMerge(richServerMaster)
                RemoveHandler richServerMaster.CalculateDocumentVariable, AddressOf richServerMaster_CalculateDocumentVariable
                ' Return the document to insert.
                e.Value = richServerMaster
                ' Required to use e.Value. Otherwise it will be ignored.
                e.Handled = True
            End If
        End Sub

'#End Region  ' #secondstage
'#Region "#thirdstage"
        ' Third stage. For each Product ID create a detailed document that will be inserted in place of the DOCVARIABLE field.
        Private Sub richServerMaster_CalculateDocumentVariable(ByVal sender As Object, ByVal e As CalculateDocumentVariableEventArgs)
            Dim currentSupplierID As Integer = GetID(e.Arguments(0).Value)
            If currentSupplierID = -1 Then Return
            If supplierID <> currentSupplierID Then
                ' Get data source that contains products for the specified supplier.
                dataDetailedForProducts = GetProductsDataFilteredbySupplier(currentSupplierID)
                supplierID = currentSupplierID
            End If

            If Equals(e.VariableName, "Product") Then
                ' Create a text engine to process a document after the mail merge.
                Dim richServerDetail As RichEditDocumentServer = New RichEditDocumentServer()
                ' Specify data source for mail merge.
                Dim options As MailMergeOptions = productRichEdit.CreateMailMergeOptions()
                options.DataSource = dataDetailedForProducts
                ' Specify that the resulting table should be joined with the header table.
                ' Do not specify this option if calculated fields are not within table cells.
                options.MergeMode = MergeMode.JoinTables
                ' Provide a procedure for further processing.
                AddHandler richServerDetail.CalculateDocumentVariable, AddressOf richServerDetail_CalculateDocumentVariable
                ' Create a merged document using the Product template. The document will contain DOCVARIABLE fields with OrderID arguments. 
                ' The CalculateDocumentVariable event for the richServerDetail fires.
                productRichEdit.MailMerge(options, richServerDetail)
                RemoveHandler richServerDetail.CalculateDocumentVariable, AddressOf richServerDetail_CalculateDocumentVariable
                ' Return the document to insert.
                e.Value = richServerDetail
                ' This setting is required for inserting e.Value into the source document. Otherwise it will be ignored.
                e.Handled = True
            End If
        End Sub

'#End Region  ' #thirdstage
'#Region "#fourthstage"
        ' Fourth stage. For each Order ID create a detailed document that will be inserted in place of the DOCVARIABLE field.
        ' This is the final stage and the Product.Orders template does not contain DOCVARIABLE fields. So, further processing is not required.
        Private Sub richServerDetail_CalculateDocumentVariable(ByVal sender As Object, ByVal e As CalculateDocumentVariableEventArgs)
            Dim currentProductID As Integer = GetID(e.Arguments(0).Value)
            If currentProductID = -1 Then Return
            If productID <> currentProductID Then
                ' Get data source that contains orders for the specified product. 
                ' The data source is obtained from the data already filtered by supplier.
                dataDetailedForOrders = GetOrderDataFilteredbyProductAndSupplier(currentProductID)
                productID = currentProductID
            End If

            If Equals(e.VariableName, "OrderDetails") Then
                Dim richServerDetailDetail As RichEditDocumentServer = New RichEditDocumentServer()
                Dim options As MailMergeOptions = ordersRichEdit.CreateMailMergeOptions()
                options.DataSource = dataDetailedForOrders
                options.MergeMode = MergeMode.JoinTables
                ordersRichEdit.MailMerge(options, richServerDetailDetail)
                e.Value = richServerDetailDetail
                e.Handled = True
            End If
        End Sub

'#End Region  ' #fourthstage
'#Region "Helper Methods"
        Protected Friend Overridable Function GetProductsDataFilteredbySupplier(ByVal supplierID As Integer) As ProductCollection
            Dim products As ProductCollection = New ProductCollection()
            For Each s As Supplier In ds
                If s.SupplierID = supplierID Then
                    products.AddRange(s.Products)
                End If
            Next

            Return products
        End Function

        Protected Friend Overridable Function GetOrderDataFilteredbyProductAndSupplier(ByVal productID As Integer) As OrderDetailCollection
            Dim orders As OrderDetailCollection = New OrderDetailCollection()
            For Each p As Product In dataDetailedForProducts
                If p.ProductID = productID Then
                    orders.AddRange(p.OrderDetails)
                End If
            Next

            Return orders
        End Function

        Protected Friend Overridable Function GetID(ByVal value As String) As Integer
            Dim result As Integer
            If Integer.TryParse(value, result) Then Return result
            Return -1
        End Function
'#End Region  ' Helper Methods
    End Class
End Namespace

File: CS/MergeProcessor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraRichEdit;
using DevExpress.XtraRichEdit.API.Native;

namespace MasterDetailExample
{
    class MergeProcessor
    {

        RichEditDocumentServer mainRichEdit = new RichEditDocumentServer();
        RichEditDocumentServer suppllierRichEdit = new RichEditDocumentServer();
        RichEditDocumentServer productRichEdit = new RichEditDocumentServer();
        RichEditDocumentServer ordersRichEdit = new RichEditDocumentServer();
        RichEditDocumentServer resultRichEdit = new RichEditDocumentServer();        
        
        SupplierCollection ds;
        ProductCollection dataDetailedForProducts;
        OrderDetailCollection dataDetailedForOrders;
        static List<int> fakeDataSource = DataHelper.CreateFakeDataSource();
        int supplierID = -1;
        int productID = -1;

        public MergeProcessor()
        {
           // Subscribe to the CalculateDocumentVariable event that triggers the master-detail report generation
            resultRichEdit.CalculateDocumentVariable += new CalculateDocumentVariableEventHandler(resultRichEdit_CalculateDocumentVariable);

            // Load main template
            mainRichEdit.LoadDocument("main.rtf");

            // Create project's data source
            ds = DataHelper.CreateData();

            // Load templates and specify data sources for RichEdit servers.

            suppllierRichEdit.LoadDocument("supplier.rtf");
            suppllierRichEdit.Options.MailMerge.DataSource = ds;
            
            productRichEdit.LoadDocument("detail.rtf");
            productRichEdit.Options.MailMerge.DataSource = ds;
            productRichEdit.Options.MailMerge.DataMember = "Products";
            
            ordersRichEdit.LoadDocument("detaildetail.rtf");
            ordersRichEdit.Options.MailMerge.DataSource = ds;
            ordersRichEdit.Options.MailMerge.DataMember = "Products.OrderDetails";

        }

        #region #startmailmerge
        // Start the process by merging main template into the document contained within the resultRichEdit server.
        public void Start()
        {
            // Since the main template contains no merge fields requiring no merge data, provide a fake data source.
            // Otherwise mailmerge will not start.
            mainRichEdit.Options.MailMerge.DataSource = fakeDataSource;
            // Trigger the multistage process. After the first mailmerge the CalculateDocumentVariable event
            //for the resultRichEdit server fires.
            mainRichEdit.MailMerge(resultRichEdit.Document);
            resultRichEdit.SaveDocument("result.docx", DocumentFormat.OpenXml);
        }
        #endregion #startmailmerge

        #region #secondstage
        // Second stage. For each Supplier ID create a detailed document that will be inserted in place of the DOCVARIABLE field.
        void resultRichEdit_CalculateDocumentVariable(object sender, CalculateDocumentVariableEventArgs e)
        {
            if (e.VariableName == "Supplier") {
                // Create a text engine to process a document after the mail merge.
                RichEditDocumentServer richServerMaster = new RichEditDocumentServer();
                // Provide a procedure for further processing
                richServerMaster.CalculateDocumentVariable += richServerMaster_CalculateDocumentVariable;
                // Create a merged document using the Supplier template. The document will contain DOCVARIABLE fields with ProductID arguments. 
                // The CalculateDocumentVariable event for the richServerMaster fires.
                suppllierRichEdit.MailMerge(richServerMaster);
                richServerMaster.CalculateDocumentVariable -= richServerMaster_CalculateDocumentVariable;                
                // Return the document to insert.
                e.Value = richServerMaster;
                // Required to use e.Value. Otherwise it will be ignored.
                e.Handled = true;
            }
        }
        #endregion #secondstage
        #region #thirdstage
        // Third stage. For each Product ID create a detailed document that will be inserted in place of the DOCVARIABLE field.
        void richServerMaster_CalculateDocumentVariable(object sender, CalculateDocumentVariableEventArgs e)
        {
            int currentSupplierID = GetID(e.Arguments[0].Value);
            if (currentSupplierID == -1)
                return;

            if (supplierID != currentSupplierID) {
                // Get data source that contains products for the specified supplier.
                dataDetailedForProducts = GetProductsDataFilteredbySupplier(currentSupplierID);
                supplierID = currentSupplierID;
            }

            if (e.VariableName == "Product") {
                // Create a text engine to process a document after the mail merge.
                RichEditDocumentServer richServerDetail = new RichEditDocumentServer();
                // Specify data source for mail merge.
                MailMergeOptions options = productRichEdit.CreateMailMergeOptions();
                options.DataSource = dataDetailedForProducts;
                // Specify that the resulting table should be joined with the header table.
                // Do not specify this option if calculated fields are not within table cells.
                options.MergeMode = MergeMode.JoinTables;
                // Provide a procedure for further processing.
                richServerDetail.CalculateDocumentVariable += richServerDetail_CalculateDocumentVariable;
                // Create a merged document using the Product template. The document will contain DOCVARIABLE fields with OrderID arguments. 
                // The CalculateDocumentVariable event for the richServerDetail fires.
                productRichEdit.MailMerge(options, richServerDetail);
                richServerDetail.CalculateDocumentVariable -= richServerDetail_CalculateDocumentVariable;
                // Return the document to insert.
                e.Value = richServerDetail;
                // This setting is required for inserting e.Value into the source document. Otherwise it will be ignored.
                e.Handled = true;
            }
        }
        #endregion #thirdstage
        #region #fourthstage
        // Fourth stage. For each Order ID create a detailed document that will be inserted in place of the DOCVARIABLE field.
        // This is the final stage and the Product.Orders template does not contain DOCVARIABLE fields. So, further processing is not required.
        void richServerDetail_CalculateDocumentVariable(object sender, CalculateDocumentVariableEventArgs e)
        {
            int currentProductID = GetID(e.Arguments[0].Value);
            if (currentProductID == -1)
                return;

            if (productID != currentProductID) {
                // Get data source that contains orders for the specified product. 
                // The data source is obtained from the data already filtered by supplier.
                dataDetailedForOrders = GetOrderDataFilteredbyProductAndSupplier(currentProductID);
                productID = currentProductID;
            }
            
            if (e.VariableName == "OrderDetails") {

                RichEditDocumentServer richServerDetailDetail = new RichEditDocumentServer();
                MailMergeOptions options = ordersRichEdit.CreateMailMergeOptions();
                options.DataSource = dataDetailedForOrders;
                options.MergeMode = MergeMode.JoinTables;
                ordersRichEdit.MailMerge(options, richServerDetailDetail);
                e.Value = richServerDetailDetail;
                e.Handled = true;
            }
        }
        #endregion #fourthstage
        #region Helper Methods
        protected internal virtual ProductCollection GetProductsDataFilteredbySupplier(int supplierID)
        {
            ProductCollection products = new ProductCollection();

            foreach (Supplier s in ds) {
                if (s.SupplierID == supplierID) {
                    products.AddRange(s.Products);
                }
            }

            return products;
        }

        protected internal virtual OrderDetailCollection GetOrderDataFilteredbyProductAndSupplier(int productID)
        {
            OrderDetailCollection orders = new OrderDetailCollection();

            foreach (Product p in dataDetailedForProducts) {
                if (p.ProductID == productID) {
                    orders.AddRange(p.OrderDetails);
                }
            }
            return orders;
        }
        protected internal virtual int GetID(string value)
        {
            int result;
            if (Int32.TryParse(value, out result))
                return result;
            return -1;
        }
        #endregion Helper Methods
    

    }
}
